barretenberg/cpp/src/barretenberg/vm2/simulation/ecc.cpp

```diff
@@ -9,7 +12,34 @@ AffinePoint Ecc::add(const AffinePoint& p, const AffinePoint& q)
     assert(q.on_curve() && "Point q is not on the curve");
 
     AffinePoint result = p + q;
-    events.emit({ .p = p, .q = q, .result = result });
+    add_events.emit({ .p = p, .q = q, .result = result });
+    return result;
+}
+
+AffinePoint Ecc::scalar_mul(const AffinePoint& point, const FF& scalar)
+{
+    uint256_t scalar_integer = static_cast<uint256_t>(scalar);
+    auto intermediate_states = std::vector<ScalarMulIntermediateState>(254);
+
+    // First iteration does conditional assignment instead of addition
+    AffinePoint temp = point;
+    // TODO(Alvaro): Call to radix here when it is implemented.
+    bool bit = scalar_integer.get_bit(0);
+
+    AffinePoint result = bit ? temp : AffinePoint::infinity();
+    intermediate_states[0] = { result, temp, bit };
+
+    for (size_t i = 1; i < 254; i++) {
+        bit = scalar_integer.get_bit(i);
+        temp = add(temp, temp);
+
+        if (bit) {
+            result = add(result, temp);
+        }
+        intermediate_states[i] = { result, temp, bit };
+    }
+    scalar_mul_events.emit(
+        { .point = point, .scalar = scalar, .intermediate_states = intermediate_states, .result = result });
```

Comment:
since intermediate states is a vector, use `.intermediate_states = std::move(intermediate_states)` then it just moves 1 pointer. Alternatively, we've been using std::arrays for these things since they are fixed size, but I think in that case it will perform a full copy (but initialization is cheaper because it's in the stack).